<a href="https://colab.research.google.com/github/rajeshmore1/DataScience_Mentorship/blob/main/17_Python_Basics_Pandas_DateTime.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Working with date and time using Pandas

In [1]:
# Code #1: Create a dates dataframe 

import pandas as pd

# Create dates dataframe with frequency
data = pd.date_range('1/1/2011', periods = 10, freq ='H')

data

DatetimeIndex(['2011-01-01 00:00:00', '2011-01-01 01:00:00',
               '2011-01-01 02:00:00', '2011-01-01 03:00:00',
               '2011-01-01 04:00:00', '2011-01-01 05:00:00',
               '2011-01-01 06:00:00', '2011-01-01 07:00:00',
               '2011-01-01 08:00:00', '2011-01-01 09:00:00'],
              dtype='datetime64[ns]', freq='H')

Datetime features can be divided into two categories. The first one time moments in a period and second the time passed since a particular period. These features can be very useful to understand the patterns in the data.

Divide a given date into features – 

pandas.Series.dt.year returns the year of the date time. 

pandas.Series.dt.month returns the month of the date time. 

pandas.Series.dt.day returns the day of the date time. 

pandas.Series.dt.hour returns the hour of the date time. 

pandas.Series.dt.minute returns the minute of the date time.

In [2]:
# Create date and time with dataframe
rng = pd.DataFrame()
rng['date'] = pd.date_range('1/1/2011', periods = 72, freq ='H')

# Print the dates in dd-mm-yy format
rng[:5]

# Create features for year, month, day, hour, and minute
rng['year'] = rng['date'].dt.year
rng['month'] = rng['date'].dt.month
rng['day'] = rng['date'].dt.day
rng['hour'] = rng['date'].dt.hour
rng['minute'] = rng['date'].dt.minute

# Print the dates divided into features
rng.head(3)


date  year  month  day  hour  minute
0 2011-01-01 00:00:00  2011      1    1     0       0
1 2011-01-01 01:00:00  2011      1    1     1       0
2 2011-01-01 02:00:00  2011      1    1     2       0

In [5]:
"""Code #4: To get the present time"""
from datetime import date

today = date.today()
print("Today's date:", today)

Today's date: 2022-05-08


In [6]:
#Example 2: Current date in different formats

from datetime import date

today = date.today()

# dd/mm/YY
d1 = today.strftime("%d/%m/%Y")
print("d1 =", d1)

# Textual month, day and year	
d2 = today.strftime("%B %d, %Y")
print("d2 =", d2)

# mm/dd/y
d3 = today.strftime("%m/%d/%y")
print("d3 =", d3)

# Month abbreviation, day and year	
d4 = today.strftime("%b-%d-%Y")
print("d4 =", d4)


d1 = 08/05/2022
d2 = May 08, 2022
d3 = 05/08/22
d4 = May-08-2022


## Timedelta
So far, we have seen how to create a DateTime object and how to format it. But sometimes, you might have to find the duration between two dates, which can be another very useful feature that you can derive from a dataset. This duration is, however, returned as a timedelta object.

In [17]:
# timedelta : duration between dates
from datetime import datetime, time

d1 = datetime(2020,4,23,11,13,10)
d2 = datetime(2021,4,23,12,13,10)
print(d1)
print(d2)
duration = d2-d1
print(type(duration))
duration

2020-04-23 11:13:10
2021-04-23 12:13:10
<class 'datetime.timedelta'>


datetime.timedelta(days=365, seconds=3600)

# Datetime and Pandas Dataframe

In [9]:
import pandas as pd

url = 'http://bit.ly/uforeports'

# read csv file
df = pd.read_csv(url)		
df.head()


City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   City             18216 non-null  object
 1   Colors Reported  2882 non-null   object
 2   Shape Reported   15597 non-null  object
 3   State            18241 non-null  object
 4   Time             18241 non-null  object
dtypes: object(5)
memory usage: 712.7+ KB


In [11]:
# Convert the Time column to datetime format
df['Time'] = pd.to_datetime(df.Time)

df.head()


City Colors Reported Shape Reported State  \
0                Ithaca             NaN       TRIANGLE    NY   
1           Willingboro             NaN          OTHER    NJ   
2               Holyoke             NaN           OVAL    CO   
3               Abilene             NaN           DISK    KS   
4  New York Worlds Fair             NaN          LIGHT    NY   

                 Time  
0 1930-06-01 22:00:00  
1 1930-06-30 20:00:00  
2 1931-02-15 14:00:00  
3 1931-06-01 13:00:00  
4 1933-04-18 19:00:00

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18241 entries, 0 to 18240
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   City             18216 non-null  object        
 1   Colors Reported  2882 non-null   object        
 2   Shape Reported   15597 non-null  object        
 3   State            18241 non-null  object        
 4   Time             18241 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(4)
memory usage: 712.7+ KB


# Day first format
By default, to_datetime() will parse string with month first (MM/DD, MM DD, or MM-DD) format, and this arrangement is relatively unique in the United State.

In most of the rest of the world, the day is written first (DD/MM, DD MM, or DD-MM). If you would like Pandas to consider day first instead of month, you can set the argument dayfirst to True.

In [18]:
df = pd.DataFrame({'date': ['3/10/2000', '3/11/2000', '3/12/2000'],
                   'value': [2, 3, 4]})
df['date'] = pd.to_datetime(df['date'], dayfirst=True)
df

date  value
0 2000-10-03      2
1 2000-11-03      3
2 2000-12-03      4

# Custome format
By default, strings are parsed using the Pandas built-in parser from dateutil.parser.parse. Sometimes, your strings might be in a custom format, for example, YYYY-DD-MM HH:MM:SS. Pandas to_datetime() has an argument called format that allows you to pass a custom format:

In [19]:
df = pd.DataFrame({'date': ['2016-6-10 20:30:0', 
                            '2016-7-1 19:45:30', 
                            '2013-10-12 4:5:1'],
                   'value': [2, 3, 4]})
print(df)
print("*****************************")
df['date'] = pd.to_datetime(df['date'], format="%Y-%d-%m %H:%M:%S")
df

                date  value
0  2016-6-10 20:30:0      2
1  2016-7-1 19:45:30      3
2   2013-10-12 4:5:1      4
*****************************


date  value
0 2016-10-06 20:30:00      2
1 2016-01-07 19:45:30      3
2 2013-12-10 04:05:01      4

## Pandas Timestamp.replace

Pandas Timestamp.replace() function is used to replace the member values of the given Timestamp. The function implements datetime.replace, and it also handles nanoseconds.

In [20]:
# Example #1: Use Timestamp.replace() function to replace the year value in the given Timestamp.

# importing pandas as pd
import pandas as pd

# Create the Timestamp object
ts = pd.Timestamp(year = 2011, month = 11, day = 21,
				hour = 10, second = 49, tz = 'US/Central')

# Print the Timestamp object
print(ts)


2011-11-21 10:00:49-06:00


In [21]:
# Now we will use the Timestamp.replace() function to replace the current year in the object with 2019.

# replace year
ts.replace(year = 2019)


Timestamp('2019-11-21 10:00:49-0600', tz='US/Central')

In [22]:
#Example #2: Use Timestamp.replace() function to replace the year, month and hour value in the given Timestamp.

# importing pandas as pd
import pandas as pd

# Create the Timestamp object
ts = pd.Timestamp(year = 2009, month = 5, day = 31,
				hour = 4, second = 49, tz = 'Europe/Berlin')

# Print the Timestamp object
print(ts)


2009-05-31 04:00:49+02:00


In [23]:
# Now we will use the Timestamp.replace() function to replace the current year, month and hour value in the object.

# replace year, month and hour value
ts.replace(year = 2019, month = 12, hour = 1)


Timestamp('2019-12-31 01:00:49+0100', tz='Europe/Berlin')

# pandas.date_range() method

pandas.date_range() is one of the general functions in Pandas which is used to return a fixed frequency DatetimeIndex.

In [24]:
# importing pandas as pd
import pandas as pd

per1 = pd.date_range(start ='1-1-2018',
		end ='1-05-2018', freq ='5H')

for val in per1:
	print(val)


2018-01-01 00:00:00
2018-01-01 05:00:00
2018-01-01 10:00:00
2018-01-01 15:00:00
2018-01-01 20:00:00
2018-01-02 01:00:00
2018-01-02 06:00:00
2018-01-02 11:00:00
2018-01-02 16:00:00
2018-01-02 21:00:00
2018-01-03 02:00:00
2018-01-03 07:00:00
2018-01-03 12:00:00
2018-01-03 17:00:00
2018-01-03 22:00:00
2018-01-04 03:00:00
2018-01-04 08:00:00
2018-01-04 13:00:00
2018-01-04 18:00:00
2018-01-04 23:00:00


In [27]:
# importing pandas as pd
import pandas as pd

dRan1 = pd.date_range(start ='1-1-2018',
		end ='8-01-2018', freq ='M')

dRan2 = pd.date_range(start ='1-1-2018',
		end ='11-01-2018', freq ='3M')

print(dRan1, '\n\n', dRan2)


DatetimeIndex(['2018-01-31', '2018-02-28', '2018-03-31', '2018-04-30',
               '2018-05-31', '2018-06-30', '2018-07-31'],
              dtype='datetime64[ns]', freq='M') 

 DatetimeIndex(['2018-01-31', '2018-04-30', '2018-07-31', '2018-10-31'], dtype='datetime64[ns]', freq='3M')


# Usecase

In [28]:
url='https://raw.githubusercontent.com/jenfly/opsd/master/opsd_germany_daily.csv'
data = pd.read_csv(url,sep=",")

In [29]:
data.head()

Date  Consumption  Wind  Solar  Wind+Solar
0  2006-01-01     1069.184   NaN    NaN         NaN
1  2006-01-02     1380.521   NaN    NaN         NaN
2  2006-01-03     1442.533   NaN    NaN         NaN
3  2006-01-04     1457.217   NaN    NaN         NaN
4  2006-01-05     1477.131   NaN    NaN         NaN

## Converting data to correct format
when working with time series, it becomes much easier if we have the Datecolumn represented as a Timestamp. Timestamp is the main pandas data structures for working with dates and times. The pandas function to_datetime() can help us convert a string to a proper date/time format.

In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4383 entries, 0 to 4382
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         4383 non-null   object 
 1   Consumption  4383 non-null   float64
 2   Wind         2920 non-null   float64
 3   Solar        2188 non-null   float64
 4   Wind+Solar   2187 non-null   float64
dtypes: float64(4), object(1)
memory usage: 171.3+ KB


In [31]:
# to explicitly convert the date column to type DATETIME
data['Date'] = pd.to_datetime(data['Date'])
data.dtypes

Date           datetime64[ns]
Consumption           float64
Wind                  float64
Solar                 float64
Wind+Solar            float64
dtype: object

We will now go ahead and set this column as the index for the dataframe using the set_index() call.

In [32]:
data = data.set_index('Date')
data

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

In [33]:
data.index # checking index of data


DatetimeIndex(['2006-01-01', '2006-01-02', '2006-01-03', '2006-01-04',
               '2006-01-05', '2006-01-06', '2006-01-07', '2006-01-08',
               '2006-01-09', '2006-01-10',
               ...
               '2017-12-22', '2017-12-23', '2017-12-24', '2017-12-25',
               '2017-12-26', '2017-12-27', '2017-12-28', '2017-12-29',
               '2017-12-30', '2017-12-31'],
              dtype='datetime64[ns]', name='Date', length=4383, freq=None)

An interesting thing to note here is thatfreq = None. What this means is that it is not known if data is collected by the hour, by day, my minute, etc. However, just by eyeballing the indices, we can see that it looks like the data was collected by the day. It would be nice to explicitly put this info into the data frame as well and we will be seeing how to do this shortly! But first a quick detour…

### What if I want to set Date and Time, both as the index?
You will come across datasets where the Date and Time were recorded as separate columns at the time of data collection. A simple yet neat trick to set them as data index is:

Concatenate the two columns but with a space between them.

The space is important!!

Convert this concatenated column to a Timestamp using to_datetime().

Set this concatenated column as the index using set_index().

In [ ]:
df['Datetime'] = pd.to_datetime(df['date'] + ' ' + df['time'])
df = df.set_index('Datetime')

## Missing value imputations
As we mentioned previously, quick glance at the data suggests it was collected at an interval of 24 hours (or a day). However, the data reflects freq = None. We can correct it as follows:

In [35]:
data_freq = data.asfreq('D')
data_freq

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

Note: Available frequencies in pandas include hourly (‘H’), calendar daily (‘D’), business daily (‘B’), weekly (‘W’), monthly (‘M’), quarterly (‘Q’), annual (‘A’), and many others.

What we have done above is say “hey, the data was collected at a day’s interval and hence each row is a new day”. So technically, we should have values for power consumption, solar production, wind production, etc for all days from 2006–2017.

In case, some days are missing in our data, the above code is going to insert empty rows and in each of these rows the values corresponding to the columns would be allNaNs. In order to avoid these NaN values, we can tell as.freq function how to fill these null values

In [36]:
data_freq = data.asfreq('D', method = 'ffill')
data_freq

Consumption     Wind   Solar  Wind+Solar
Date                                                
2006-01-01   1069.18400      NaN     NaN         NaN
2006-01-02   1380.52100      NaN     NaN         NaN
2006-01-03   1442.53300      NaN     NaN         NaN
2006-01-04   1457.21700      NaN     NaN         NaN
2006-01-05   1477.13100      NaN     NaN         NaN
...                 ...      ...     ...         ...
2017-12-27   1263.94091  394.507  16.530     411.037
2017-12-28   1299.86398  506.424  14.162     520.586
2017-12-29   1295.08753  584.277  29.854     614.131
2017-12-30   1215.44897  721.247   7.467     728.714
2017-12-31   1107.11488  721.176  19.980     741.156

[4383 rows x 4 columns]

ffill refers to forward fill. This means when a null is encountered for a particular column, it will be replaced by the value in the previous row. As an example:

## Resampling
Resampling simply refers to an aggregation of data over a certain time period. Its performance is similar to the group by function in SQL, i.e. data is first split into time bins and some computation is performed on each bin. For instance, given our daily data, we can resample (or bin) it by the month or year and generate some relevant statistics such as minimum, maximum, or mean values.

### Weekly Resampling
To compute weekly mean values for electricity consumption, wind, and solar production:



In [37]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_weekly_mean = data[data_columns].resample('W').mean() # W stands for weekly
data_weekly_mean


Consumption        Wind      Solar  Wind+Solar
Date                                                      
2006-01-01  1069.184000         NaN        NaN         NaN
2006-01-08  1381.300143         NaN        NaN         NaN
2006-01-15  1486.730286         NaN        NaN         NaN
2006-01-22  1490.031143         NaN        NaN         NaN
2006-01-29  1514.176857         NaN        NaN         NaN
...                 ...         ...        ...         ...
2017-12-03  1536.236314  284.334286  18.320857  302.655143
2017-12-10  1554.824946  636.514714  16.440286  652.955000
2017-12-17  1543.856889  442.531857  18.143714  460.675571
2017-12-24  1440.342401  339.018429   9.895143  348.913571
2017-12-31  1203.265211  604.699143  19.240143  623.939286

[627 rows x 4 columns]

If you notice the indices in the output, you would observe that they are at a gap of one week, we have Jan1 2006 followed by Jan8 2006, and so on. The first row above, labeled 2006–01–01, contains the mean of all the data contained in the time bin 2006–01–01 through 2006–01–07.

###Monthly Resampling
Similarly, to compute monthly maximum values for electricity consumption, wind, and solar production:



In [38]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_monthly_max = data[data_columns].resample('M').max() # W stands for weekly
data_monthly_max

Consumption     Wind    Solar  Wind+Solar
Date                                                 
2006-01-31   1613.31200      NaN      NaN         NaN
2006-02-28   1606.66700      NaN      NaN         NaN
2006-03-31   1545.83400      NaN      NaN         NaN
2006-04-30   1447.79800      NaN      NaN         NaN
2006-05-31   1380.17500      NaN      NaN         NaN
...                 ...      ...      ...         ...
2017-08-31   1434.05200  455.379  216.117     601.010
2017-09-30   1460.61621  761.751  168.123     838.572
2017-10-31   1508.51484  819.818  123.715     846.617
2017-11-30   1617.03309  743.280   57.358     772.775
2017-12-31   1651.90418  812.422   42.994     822.371

[144 rows x 4 columns]

##Rolling windows
This is quite similar to the resampling process that we just learned. The difference is that the bins over which some aggregating functions are performed) are overlapping.

Bins in case of weekly resampling: Jan1- Jan 7; Jan8 - Jan14, Jan 15 - Jan 21, etc

Bins in case of weekly rolling: Jan1- Jan7; Jan 2- Jan 8, Jan 3- Jan 9, etc.

To compute a 7-day rolling mean:

In [39]:
data_columns = ['Consumption', 'Wind', 'Solar', 'Wind+Solar']
data_7d_rol = data[data_columns].rolling(window = 7, center = True).mean()
data_7d_rol


Consumption        Wind      Solar  Wind+Solar
Date                                                      
2006-01-01          NaN         NaN        NaN         NaN
2006-01-02          NaN         NaN        NaN         NaN
2006-01-03          NaN         NaN        NaN         NaN
2006-01-04  1361.471429         NaN        NaN         NaN
2006-01-05  1381.300143         NaN        NaN         NaN
...                 ...         ...        ...         ...
2017-12-27  1208.214129  617.734286  17.807143  635.541429
2017-12-28  1203.265211  604.699143  19.240143  623.939286
2017-12-29          NaN         NaN        NaN         NaN
2017-12-30          NaN         NaN        NaN         NaN
2017-12-31          NaN         NaN        NaN         NaN

[4383 rows x 4 columns]

In the above command, center = True means for the time bin, say Jan 1 to Jan 8, the rolling mean would be calculated and placed next to the center of the bin i.e. Jan 4. To make it clearer, let's check the output of the above code:

As you can see, the mean consumption value (of 1361.471) in 2006–01–04 was calculated by averaging the values from 2006–01–01 to 2006–01–07. Similarly, the mean consumption value (of 1381.300) in 2006–01–05 was calculated by averaging the values from 2006–01–02 to 2006–01–08.

By the same logic, to calculate the mean consumption value for 2006–01–01, we would need to average the values from 2005–12–29 to 2006–01–04. However, the data for the year 2005 is missing and thus we obtain a Null for the first few rows.


For more info: https://towardsdatascience.com/time-series-analysis-using-pandas-in-python-f726d87a97d8